elixir/apps/web/test/web/live/settings/identity_providers/jumpcloud/connect_test.exs

```diff
@@ -0,0 +1,303 @@
+        conn
+        |> authorize_conn(identity)
+        |> assign(:account, account)
+        |> get(
+          ~p"/#{account.id}/settings/identity_providers/jumpcloud/#{provider.id}/redirect",
+          %{}
+        )
+
+      assert to = redirected_to(conn)
+      uri = URI.parse(to)
+      assert uri.host == "localhost"
+      assert uri.path == "/authorize"
+
+      callback_url =
+        url(
+          ~p"/#{account.id}/settings/identity_providers/jumpcloud/#{provider.id}/handle_callback"
+        )
+
+      {_params, state, verifier} =
+        conn.cookies["fz_auth_state_#{provider.id}"]
+        |> :erlang.binary_to_term()
+
+      code_challenge = Domain.Auth.Adapters.OpenIDConnect.PKCE.code_challenge(verifier)
+      scope_string = ~w[
+       openid
+       email
+       profile
+     ] |> Enum.join(" ")
+
+      assert URI.decode_query(uri.query) == %{
+               "access_type" => "offline",
+               "client_id" => provider.adapter_config["client_id"],
+               "code_challenge" => code_challenge,
+               "code_challenge_method" => "S256",
+               "redirect_uri" => callback_url,
+               "response_type" => "code",
+               "scope" => scope_string,
+               "state" => state
+             }
+    end
+  end
+
+  describe "handle_idp_callback/2" do
+    setup do
+      account = Fixtures.Accounts.create_account()
+      %{account: account}
+    end
+
+    test "redirects to login page when user is not signed in", %{conn: conn} do
+      account_id = Ecto.UUID.generate()
+      provider_id = Ecto.UUID.generate()
+
+      conn =
+        conn
+        |> get(~p"/#{account_id}/sign_in/providers/#{provider_id}/handle_callback", %{
+          "state" => "foo",
+          "code" => "bar"
+        })
+
+      assert redirected_to(conn) == "/#{account_id}"
+      assert flash(conn, :error) == "Your session has expired, please try again."
+    end
+
+    test "redirects with an error when state cookie does not exist", %{
+      account: account,
+      conn: conn
+    } do
+      {provider, _bypass} =
+        Fixtures.Auth.start_and_create_jumpcloud_provider(account: account)
+
+      actor = Fixtures.Actors.create_actor(type: :account_admin_user, account: account)
+      identity = Fixtures.Auth.create_identity(account: account, actor: actor, provider: provider)
+
+      conn =
+        conn
+        |> authorize_conn(identity)
+        |> assign(:account, account)
+        |> get(
+          ~p"/#{account}/settings/identity_providers/jumpcloud/#{provider}/handle_callback",
+          %{
+            "state" => "XOXOX",
+            "code" => "bar"
+          }
+        )
+
+      assert redirected_to(conn) ==
+               ~p"/#{account}/settings/identity_providers/jumpcloud/#{provider}"
+
+      assert flash(conn, :error) == "Your session has expired, please try again."
+    end
+
+    ## TODO: Re-work this test since the API key is hard coded/long lived for JumpCloud API
```

Comment:
I just realized I forgot to update this test.  I'll update this, but I think the rest of the PR should be ready for review.

Reply:
Updated.